### DISCUSSION TITLE
Can we incorporate stateful (per ordered partition) Custom Filter function to local Clickhouse server?

### DISCUSSION THREAD

**Discussion by unjeru** [ClickHouse/ClickHouse#53446] [Q&A]
I implement a simple logic based on Counting Bloom filter (Java) that allows to get/skip first N rows per any combination of fields (groups cardinality ~ count/20) for query response.

This approach has several disadvantages:
- I can't add filtering in SQL query, it's some distinct logic.
- it creates big enough network load
- I can't use final aggregate functions on server (Clickhouse) side side - so I have to implement aggregate functions on backend service.

The final idea to add single-thread function with one state per partition where with arguments like 
```
<(calculable) column>  - expression 
<partitioned by> - column with partitions 
<threshold> - uint8
<cbf atttributes ...> - vector size, number of hashes
```
So it could look like
```
select * from event_log where approxGetFirstN(producer, zoneId, 5, magic .. numbers) # get first five
select * from event_log where not approxGetFirstN(producer, zoneId, 5, magic .. numbers) # skip first five
```

Or It's still much better that now if I can execute N queries for each available partition.

**Comment by UnamedRus** [comment]
What are you trying to implement?

Ability to not read at least 1 granule from all part(itions) for simple query

```
SELECT * FROM table LIMIT 5;
SELECT * FROM table ORDER BY event_time LIMIT 5; -- where partition by is toYYYYMM(event_time)
```

?

**Reply by unjeru** [reply]
In my case an approach above requires ~100M queries per one "business-valuable" response, because there are a lot of producers with different features - `(groups cardinality ~ count/20) `
The same time we can choose the accuracy of the response using approx counting of producers events.

The idea is to fill and check all counters in one pass, and we are OK with that this will be an approx result.

**Reply by UnamedRus** [reply]
> ~100M queries per one "business-valuable" response,

That's a lot for ClickHouse not sure that it can be valuable. (theoretically you can make secondary reverse index table, which will be much smaller and before running query against main one, you will check this secondary table)

Related https://github.com/ClickHouse/ClickHouse/pull/49468

Plus there was some pr which would make one bloom filter for order by index columns, so ClickHouse could check does this value exist at all in that specific part.
